File: banchus-app/src/main/java/pe/nanamochi/banchus/protocol/PacketHandler.java
package pe.nanamochi.banchus.protocol;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;
import pe.nanamochi.banchus.core.Packet;
import pe.nanamochi.banchus.core.Packets;
import pe.nanamochi.banchus.database.entity.Session;

@Slf4j
@Component
public class PacketHandler {
  private final Map<Packets, AbstractPacketHandler<?>> handlers;

  public PacketHandler(List<AbstractPacketHandler<?>> beans) {
    this.handlers =
        beans.stream()
            .collect(Collectors.toMap(AbstractPacketHandler::getPacketType, Function.identity()));
  }

  public void handle(Packet packet, Session session, ByteArrayOutputStream responseStream)
      throws IOException {
    if (packet == null) return;

    AbstractPacketHandler<?> handler = handlers.get(packet.getPacketType());
    if (handler == null) return;

    dispatch(handler, packet, session, responseStream);
  }

  private <T extends Packet> void dispatch(
      AbstractPacketHandler<T> handler,
      Packet packet,
      Session session,
      ByteArrayOutputStream responseStream)
      throws IOException {
    log.debug("Handling packet: {}", packet.getPacketType());
    HandleClientPacket annotation = handler.getClass().getAnnotation(HandleClientPacket.class);
    boolean shouldCheck = (annotation != null) && annotation.checkForRestriction();

    if (shouldCheck && session.getUser().isRestricted()) return;

    handler.handle(handler.getPacketClass().cast(packet), session, responseStream);
  }
}

File: banchus-app/src/main/kotlin/pe/nanamochi/banchus/infrastructure/protocol/PacketHandler.kt
package pe.nanamochi.banchus.infrastructure.protocol

import java.io.ByteArrayOutputStream
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import pe.nanamochi.banchus.auth.entity.Session
import pe.nanamochi.banchus.core.ClientPacket
import pe.nanamochi.banchus.core.PacketType

@Component
class PacketHandler(beans: List<AbstractPacketHandler<*>>) {
    private val log = LoggerFactory.getLogger(javaClass)

    private val handlers: Map<PacketType, AbstractPacketHandler<*>> = beans.associateBy { it.type }

    fun handle(packet: ClientPacket, session: Session, responseStream: ByteArrayOutputStream) {
        val handler = handlers[packet.type] ?: return

        val annotation =
            handler::class.annotations.find { it is HandleClientPacket } as? HandleClientPacket
        if (annotation?.checkForRestriction == true && session.isRestricted == true) {
            return
        }

        log.debug(
            "Handling packet of type {} with handler {}",
            packet.type,
            handler::class.simpleName,
        )
        dispatch(handler, packet, session, responseStream)
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T : ClientPacket> dispatch(
        handler: AbstractPacketHandler<T>,
        packet: ClientPacket,
        session: Session,
        responseStream: ByteArrayOutputStream,
    ) {
        handler.handle(packet as T, session, responseStream)
    }
}
